src/core/router/tunnel/config.cc

```diff
@@ -32,134 +32,205 @@
 #include "core/router/i2np.h"
 
+#include "core/util/byte_stream.h"
+#include "core/util/log.h"
 #include "core/util/timestamp.h"
 
 namespace kovri {
 namespace core {
 
 // TODO(unassigned): refactor all tunnel implementation (applies across entire namespace)
 
+TunnelAESRecordAttributes::TunnelAESRecordAttributes() {
+  RandBytes(layer_key.data(), layer_key.size());
+  RandBytes(IV_key.data(), IV_key.size());
+  RandBytes(reply_key.data(), reply_key.size());
+  RandBytes(reply_IV.data(), reply_IV.size());
+}
+
 TunnelHopConfig::TunnelHopConfig(
-    std::shared_ptr<const kovri::core::RouterInfo> r) {
-  kovri::core::RandBytes(layer_key, 32);
-  kovri::core::RandBytes(iv_key, 32);
-  kovri::core::RandBytes(reply_key, 32);
-  kovri::core::RandBytes(reply_IV, 16);
-  kovri::core::RandBytes(rand_pad, 29);
-  tunnel_ID = kovri::core::Rand<std::uint32_t>();
-  is_gateway = true;
-  is_endpoint = true;
-  router = r;
-  next_router = nullptr;
-  next_tunnel_ID = 0;
-  next = nullptr;
-  prev = nullptr;
-  record_index = 0;
+    std::shared_ptr<const RouterInfo> router)
+    : m_CurrentRouter(router),
+      m_TunnelID(Rand<std::uint32_t>()),
+      m_AESRecordAttributes(),
+      m_NextRouter(nullptr),
+      m_NextTunnelID(0),
+      m_PreviousHop(nullptr),
+      m_NextHop(nullptr),
+      m_IsGateway(true),
+      m_IsEndpoint(true),
+      m_RecordIndex(0) {
+        if (!router)
+          throw std::runtime_error("TunnelHopConfig: current router is null");
+      }
+
+std::shared_ptr<const RouterInfo> TunnelHopConfig::GetCurrentRouter() const noexcept {
+  return m_CurrentRouter;
 }
 
 void TunnelHopConfig::SetNextRouter(
-    std::shared_ptr<const kovri::core::RouterInfo> r) {
-  next_router = r;
-  is_endpoint = false;
-  next_tunnel_ID = kovri::core::Rand<std::uint32_t>();
+    std::shared_ptr<const RouterInfo> router,
+    std::uint32_t tunnel_id,
+    bool is_endpoint) {
+  if (!router)
+    throw std::runtime_error("TunnelHopConfig: next router is null");
+  m_NextRouter = router;
+  m_NextTunnelID = tunnel_id;
+  m_IsEndpoint = is_endpoint;
 }
 
-void TunnelHopConfig::SetReplyHop(
-    const TunnelHopConfig* reply_first_hop) {
-  next_router = reply_first_hop->router;
-  next_tunnel_ID = reply_first_hop->tunnel_ID;
-  is_endpoint = true;
+std::shared_ptr<const RouterInfo> TunnelHopConfig::GetNextRouter() const noexcept {
+  return m_NextRouter;
 }
 
-void TunnelHopConfig::SetNext(
-    TunnelHopConfig* n) {
-  next = n;
-  if (next) {
-    next->prev = this;
-    next->is_gateway = false;
-    is_endpoint = false;
-    next_router = next->router;
-    next_tunnel_ID = next->tunnel_ID;
+void TunnelHopConfig::SetReplyHop(const TunnelHopConfig* hop) {
+  if (!hop)
+    throw std::runtime_error("TunnelHopConfig: reply hop is null");
+  SetNextRouter(hop->GetCurrentRouter(), hop->GetTunnelID(), true);
+}
+
+void TunnelHopConfig::SetNextHop(TunnelHopConfig* hop) {
+  m_NextHop = hop;
+  if (m_NextHop) {
+    m_NextHop->m_PreviousHop = this;
```

Comment:
Presumably this is why a raw ptr was chosen over a `shared_ptr` (cyclic reference) - what about a `weak_ptr`? Lifetime has to be mandated externally, which can be tough for larger projects sometimes ...

I saw how this was constructed after reading some more of the code. What if `setNextHop` took a `unique_ptr` - it would obvious that it was taking ownership of the `hop`.